ConnectionController周りで落ちないようにする

tsuu32 opened the pull request:
### やったこと
いっぱいやりました。
#### バグ修正
* 起動直後のWelcome Sceneで、[join as a Listener] ボタンを押して[Back]で戻るとDoneが押せてしまう不具合を修正（これによってDJでもリスナでもない状態になってしまうことがなくなる）
* DJとして参加し、[Reconnect]を押し、[join as a Listener] ボタンを押したが「やっぱやめた（DJを選ばずにViewを閉じる）」したとき、リスナになってしまっている不具合を修正（DJがリスナに切り替わるのはJoin as a Listener Sceneを表示した時ではなく、実際にDJをタップしたときになった）
-> join as a Listenerというボタンの名前は適切か？（正確には「周囲にいるDJを見る」機能）
* Sessionタブの子機一覧で、twitter連携してない子機のアイコンに他の人のアイコンが表示されてしまう不具合を修正 280052d
* リスナが別のDJに繋いだとき、Sessionタブの親機のアイコンに以前繋がっていたDJのアイコンが表示されてしまう不具合を修正

#### 新機能
* Join as a Listener Sceneを改良
  * 「DJ: 端末名」という味気ないものから、アイコンと名前を表示するようにした（アイコンはtwitterのものに対応）

### 技術的な話
* isDJをBool!型からBool?型にする
isDJがnilのときWelcome Sceneを閉じられないようにした。

* connectedDJをMCPeerID!型からMCPeerID?型にする
リスナ視点で、接続中の親機がいないとき（つまり、親機を見失ったとき）connectedDJにnilが入る。

* initializeはアプリ内で一回だけ呼ぶ
今まで[join as the DJ] ボタンを押したときと[join as a Listener]ボタンを押したときでinitializeし直していた(sessionを作り直していた）。よく考えるとその必要はない（session.disconnect()すればいい）ので、initializeはアプリ内で一度だけ呼ばれるようにする。

* browserがBrowsingしてるのはJoin as a Listener Sceneにいるときだけ
viewWillDisappearメソッドを使ってJoin as a Listener Sceneを離れるとき（DJを選択したときか[Back]を押したとき）にbrowseを止める（connectableDJs.removeAll()もする）。

### やってほしいこと

実機で試してみて、なんかバグが発生しないか。コードにアレなところないか。

tsuu32 committed:
Sessionタブの子機一覧で、twitter連携してない子機のアイコンに他の人のアイコンが表示されるのを防ぐ

tsuu32 committed:
ConnectionControllerDelegateの使われてない関数を削除

connectionController(didReceiveData data: Data, from peerID: MCPeerID)
消す

tsuu32 committed:
isDJをBool?型にする他でいろいろ直す

isDJをBool?型にしたことでisDJ!にしたところ多数
方針変換: いままでListenerConnectionViewを表示するたびにinitializeしていた
でも、「やっぱやめた」を考えるとまずいことが起こるので
initializeはWelcomeViewを表示したはじめの一回だけ行う

ListenerConnectionViewController.swift:
initializeしない
viewDidLoadでstartBrowse()
viewWillDisappearでstopBrowse()
表示するDJ名はconnectableDJNameCorrespondenceを使う（アイコン対応）
（変なところにあったtableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath)の位置を変える）

MemberViewController.swift
connectedDJをMCPeerID?型にしたことでちょっと変更（親機を見失った子機の動作を考慮したコード）
DJIconはtwitterのアイコンがなかったらnil（今後ランダムな画像を入れたい）

WelcomeViewController.swift:viewDidLoadでinitializeする
isDJがnilのとき閉じられないようにする

yaplus commented on DJYusaku/WelcomeViewController.swift:
```diff
@@ -17,33 +17,32 @@ class WelcomeViewController: UIViewController {
         
-        self.doneButtonItem.isEnabled = WelcomeViewController.isViewAppearedAtLeastOnce
-        WelcomeViewController.isViewAppearedAtLeastOnce = true
+        if let _ = ConnectionController.shared.isDJ {
```
ConnectionController.shared.isDJ != nilにしましょう

tsuu32 replied:
d1dba39 で修正

tsuu32 committed:
ちょっとした修正

ConnectionController.swift:
startDJメソッドが呼ばれるときにbrowserがbrowsingなことはないので削除

SearchViewController.swift:
「connectedDJがいないとき、リスナはsendRequestを行わない」という動作をif letではなくguardを用いるようにする

WelcomeViewController.swift:
いい感じにする

tsuu32 commented:
[join as a Listener]ボタンで発生する動作が、実際には「近くにいるDJを探す」動作になってるの気になる。

（DJの役割の人が[join as a Listener]ボタンを押しても、[join as a Listener]ボタン押すだけならリスナになるわけじゃないため。）

yaplus commented:
@tsuu32 

> [join as a Listener]ボタンで発生する動作が、実際には「近くにいるDJを探す」動作になってるの気になる。

https://github.com/enpitut2019/dj-yusaku/issues/34 もそうだけど画面変えるときに考えとこう

Main.storyboard のコンフリクトは解決できそう？

yaplus commented:
よく見たらブランチ名が `fix/wazap` で草

tsuu32 committed:
Merge branch 'master' into fix/wazap

tsuu32 commented:
@yaplus コンフリクト解決しました (全部`fix/wazap`に合わせた）

yaplus reviewed (approved):
ヨシ！😺👷🏾‍♂️
